### PR TITLE
Add mutator for modifying authenticationSession with external API

### PR DESCRIPTION
**Pull request by kubadz**
## Related issue

#228
@aeneasr @piotrmsc
## Proposed changes

As stated in the linked issue - this PR adds a mutator which make an upstream call to an API with AuthenticationSession as payload and expect AuthenticationSession as a response.

## Checklist

- [x] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md)
- [x] I have read the [security policy](../security/policy)
- [x] I confirm that this pull request does not address a security vulnerability. If this pull request addresses a security
vulnerability, I confirm that I got green light (please contact [security@ory.sh](mailto:security@ory.sh)) from the maintainers to push the changes.
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation within the code base (if appropriate)
- [ ] I have documented my changes in the [developer guide](https://github.com/ory/docs) (if appropriate)

### CODE REVIEW

**Comment by aeneasr** [comment]
sorry - I squash-merged the other PR causing some merge conflicts

**Comment by kubadz** [comment]
> sorry - I squash-merged the other PR causing some merge conflicts

No problem :) fortunately, there are not many conflicts

File: driver/configuration/provider_viper_mutator.go [hunk 1]
```diff
@@ -14,6 +14,8 @@ const (
 
 	ViperKeyMutatorNoopIsEnabled = "mutators.noop.enabled"
 
+	ViperKeyMutatorHydratorIsEnabled = "mutators.Hydrator.enabled"
```
Comment (jakkab) [inline driver/configuration/provider_viper_mutator.go]:
```suggestion
	ViperKeyMutatorHydratorIsEnabled = "mutators.hydrator.enabled"
```

**Comment by aeneasr** [comment]
This looks great!! All that's left is adding docs here: https://github.com/ory/docs/blob/master/docs/oathkeeper/pipeline/mutator.md :)

File: pipeline/mutate/mutator_hydrator.go [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+}
+
+type RetryConfig struct {
+	NumberOfRetries     int `json:"number"`
+	DelayInMilliseconds int `json:"delayInMilliseconds"`
+}
+
+type externalAPIConfig struct {
+	Url   string          `json:"url"`
+	Authn *Authentication `json:"authn,omitempty"`
```
Comment (aeneasr) [inline pipeline/mutate/mutator_hydrator.go]: This would probably be `authorization` right? Because we're modifying HTTP Authorization?

Comment (kubadz) [reply pipeline/mutate/mutator_hydrator.go]: Sure, we are modifying `HTTP Authorization` header, but it is a header used for authentication. And the config is actually specifying [*Authentication* scheme](https://developer.mozilla.org/en-US/docs/Web/HTTP/Authentication#Authentication_schemes) and credentials. But if this argumentation doesn't convince you and you prefer it to be `Authorization` I will not argue to change that 😄

Comment (aeneasr) [reply pipeline/mutate/mutator_hydrator.go]: We could also just change it to `auth`, then it implies both 😅

Comment (kubadz) [reply pipeline/mutate/mutator_hydrator.go]: Ok, changed it. Let me know what you think about it now 😄 
I will adjust documentation in a second

**Comment by aeneasr** [comment]
Wohoooo!

**Pull request merged**